fix(@mastra/core, @mastra/memory): Memory + useChat + client tools

TylerBarnes opened the pull request:
Fixes https://discord.com/channels/1309558646228779139/1347999511280222279/1348735064590127209

When using `useChat` it sends the full message list to the server on each request. Since memory also includes the full chat history this means the chat history would get mixed up or doubled. When pairing this with client side tool calling, incomplete tool message errors would happen because the message history included the original saved tool call (which had no separate tool result) + the new client side injected tool call which included a result - meaning the message history would include the initial tool call and then the same tool call with a result.

These issues are fixed and I updated the documentation to show how to use `useChat` with memory.

TylerBarnes committed:
filter out first tool call when there's a dupe (for client tools)

TylerBarnes committed:
don't embed empty strings

TylerBarnes committed:
support passing in a single structured message

TylerBarnes committed:
update docs for memory + useChat

TylerBarnes committed:
ch

TylerBarnes committed:
cleanup

wardpeet commented on packages/core/src/agent/index.ts:
```diff
@@ -417,6 +417,7 @@ export class Agent<
   sanitizeResponseMessages(messages: Array<CoreMessage>): Array<CoreMessage> {
     let toolResultIds: Array<string> = [];
     let toolCallIds: Array<string> = [];
+    let dupeToolCalls = new Set<string>();
```
You should move to Set instead of array

wardpeet replied:
no when i checked it it said array :D

wardpeet replied:
or maybe i was blind

TylerBarnes replied:
no it was like that and then I pushed up another commit haha

wardpeet reviewed (approved):
Added a comment but nothing blocking. Can we add tests in a follow up

TylerBarnes commented:
hmm broke some tests, gonna fix em now

TylerBarnes committed:
restore previous incomplete tool handling

TylerBarnes committed:
restore original behaviour

TylerBarnes commented:
> Added a comment but nothing blocking. Can we add tests in a follow up

@wardpeet I went to add a test and discovered this was almost entirely a documentation problem. There were a couple small fixes needed but I reverted the dupe tool checking and it still works lol. So basically there are a couple small fixes needed here: don't create embeddings from empty strings + allow passing a single structured message. Besides that it turns out the code I added just helped me figure out how this works and how to document it 😆 
I might look at automatically detecting when someone is using `useChat` incorrectly, and then add a test for that, but that'd be a different PR.
